consensus-types/types/execution_requests.go

```diff
@@ -0,0 +1,196 @@
+)
+
+// EncodedExecutionRequest is the result of GetExecutionRequestsList which is spec defined.
+type EncodedExecutionRequest = bytes.Bytes
+
+type ExecutionRequests struct {
+	Deposits       []*DepositRequest
+	Withdrawals    []*WithdrawalRequest
+	Consolidations []*ConsolidationRequest
+}
+
+func (e *ExecutionRequests) ValidateAfterDecodingSSZ() error {
+	return nil
+}
+
+// GetExecutionRequestsList introduced in pectra from the consensus spec
+// https://github.com/ethereum/consensus-specs/blob/dev/specs/electra/beacon-chain.md#new-get_execution_requests_list
+func GetExecutionRequestsList(er *ExecutionRequests) ([]EncodedExecutionRequest, error) {
+	if er == nil {
+		return nil, errors.New("nil execution requests")
+	}
+	result := make([]EncodedExecutionRequest, 0)
+
+	// Process deposit requests if non-empty.
+	if len(er.Deposits) > 0 {
+		requests := DepositRequests(er.Deposits)
+		depositBytes, err := requests.MarshalSSZ()
+		if err != nil {
+			return nil, err
+		}
+		combined := append([]byte{}, depositRequestType()...)
+		combined = append(combined, depositBytes...)
+		result = append(result, combined)
+	}
+
+	// Process withdrawal requests if non-empty.
+	if len(er.Withdrawals) > 0 {
+		requests := WithdrawalRequests(er.Withdrawals)
+		withdrawalBytes, err := requests.MarshalSSZ()
+		if err != nil {
+			return nil, err
+		}
+		combined := append([]byte{}, withdrawalRequestType()...)
+		combined = append(combined, withdrawalBytes...)
+		result = append(result, combined)
+	}
+
+	// Process consolidation requests if non-empty.
+	if len(er.Consolidations) > 0 {
+		requests := ConsolidationRequests(er.Consolidations)
+		consolidationBytes, err := requests.MarshalSSZ()
+		if err != nil {
+			return nil, err
+		}
+		combined := append([]byte{}, consolidationRequestType()...)
+		combined = append(combined, consolidationBytes...)
+		result = append(result, combined)
+	}
+
+	return result, nil
+}
+
+// DecodeExecutionRequests is used to decode the result from GetPayload into an ExecutionRequests.
+// TODO(pectra): Change this to use []EncodedExecutionRequest as input and fix tests.
+func DecodeExecutionRequests(encodedRequests [][]byte) (*ExecutionRequests, error) {
+	var result ExecutionRequests
+	var prevType *uint8
+
+	// Iterate over each encoded request group.
+	for _, encoded := range encodedRequests {
+		if len(encoded) < 1 {
+			return nil, errors.New("invalid execution request, length less than 1")
+		}
+		// The first byte indicates the request type.
+		reqType := encoded[0]
+
+		// Enforce that request types are in strictly increasing order.
+		if prevType != nil && *prevType >= reqType {
+			return nil, errors.New("requests should be in sorted order and unique")
+		}
+		prevType = &reqType
+
+		// The remaining bytes are the SSZ serialization for this group.
+		data := encoded[1:]
+
+		// Switch based on the request type.
+		switch reqType {
+		case depositRequestType()[0]:
+			req, err := DecodeDepositRequests(data)
```

Comment:
in this switch we are implicitly assuming that if `err == nil` then `req != nil`. It should be fine but I wonder if we could return the actual type rather than a pointer to it?
It would avoid the deference (and the types themselves are slices so kinda pointers already)